src/typechecker/FStar.TypeChecker.Rel.fs

```diff
@@ -654,44 +657,111 @@ let flex_uvar_head t =
     | Tm_uvar (u, _) -> u
     | _ -> failwith "Not a flex-uvar"
 
-let destruct_flex_t t wl : flex_t * worklist =
-    let head, args = U.head_and_args t in
+(* Make sure the uvar at the head of t0 is not affected by a
+ * substitution.
+ *
+ * In the case that it is, first solve it to a new appropriate uvar
+ * without a substitution. This function returns t again, though
+ * it is unchanged (the changes only happen in the UF graph).
+ *
+ * The way we generate the new uvar is find the x's in the domain of ?u
+ * that are not affected by the substitution and make a new variable with
+ * only those binders in the domain. The ones that are affected we will
+ * "hoist" by making the type of ?u' an arrow.
+ *
+ * Example:
+ * If we have (x,y |- ?u : ty)[(x:a) <- x, (y:int) <- 42], we will make
+ * ?u' with x in its binders, abstracted over y:
+ *
+ * (x |- ?u') : int -> ty
+ *
+ * and then solve
+ *
+ * ?u <- (?u' y)
+ *
+ * Which means the original term now compresses to ?u' 42. The flex
+ * problem we now return is
+ *
+ * ?u', [42]
+ *
+ * We return early if the substitution is empty or if the uvar is
+ * totally unaffected by it.
+ *
+ * (It's not strictly needed for this function to return t, compressing
+ *  t0 will give you the same term.)
+ *)
+let ensure_no_uvar_subst (t0:term) (wl:worklist)
+  : term * worklist
+  = (* Returns true iff the variable x is not affected by substitution s *)
+    let not_affected_by (s:subst_ts) (x:bv) : bool =
+      let t_x = S.bv_to_name x in
+      let t_x' = SS.subst' s t_x in
+      match (SS.compress t_x').n with
+      | Tm_name y ->
+         S.bv_eq x y // Check if substituting returned the same variable
+      | _ -> false
+    in
+    (* Splits a gamma into the part which is not affected by the substitution s
+     * and the part that is. *)
+    let split_gamma (s:subst_ts) (gamma:list<binding>) : list<binding> * list<binding> =
+      gamma |> List.partition (function
+      | Binding_var x -> not_affected_by s x
+      | _ -> true)
+    in
```

Comment:
I don't think it's safe to split gamma this way.

The gamma is telescope of binders and splitting binders to either side could make both returned lists ill-formed. E.g., consider

```
Gamma = x:tx, y: ty x, z:tz y
```
with only `y` affected by the substitution.

This seems to be a problem in the original code itself.

How about instead taking the least suffix of gamma that mentions all the affected variables.

Reply:
Good point, yes. The only reason we do not simply take all binders is for performance, right? (Checking so I can write a proper comment.)